src/compute/src/extensions/reduce.rs

```diff
@@ -16,61 +16,60 @@
 use differential_dataflow::Data;
 use timely::dataflow::Scope;
 
 use crate::extensions::arrange::ArrangementSize;
 
 /// Extension trait for the `reduce_abelian` differential dataflow method.
-pub(crate) trait MzReduce<G: Scope, K: Data, V: Data, R: Semigroup>:
-    ReduceCore<G, K, V, R>
+pub(crate) trait MzReduce<G: Scope, T1: TraceReader<Time = G::Timestamp>>
 where
     G::Timestamp: Lattice + Ord,
 {
     /// Applies `reduce` to arranged data, and returns an arrangement of output data.
     fn mz_reduce_abelian<L, T2>(&self, name: &str, logic: L) -> Arranged<G, TraceAgent<T2>>
     where
-        T2: Trace
-            + for<'a> TraceReader<Key<'a> = &'a K, KeyOwned = K, Time = G::Timestamp>
-            + 'static,
+        T2: for<'a> Trace<Key<'a> = T1::Key<'a>, Time = G::Timestamp> + 'static,
         T2::ValOwned: Data,
         T2::Diff: Abelian,
         T2::Batch: Batch,
         T2::Builder:
-            Builder<Output = T2::Batch, Item = ((T2::KeyOwned, T2::ValOwned), T2::Time, T2::Diff)>,
-        L: FnMut(&K, &[(&V, R)], &mut Vec<(T2::ValOwned, T2::Diff)>) + 'static,
-        Arranged<G, TraceAgent<T2>>: ArrangementSize,
-    {
-        // Allow access to `reduce_abelian` since we're within Mz's wrapper.
-        #[allow(clippy::disallowed_methods)]
-        self.reduce_abelian::<_, T2>(name, logic)
-            .log_arrangement_size()
-    }
+            Builder<Output = T2::Batch, Item = ((T1::KeyOwned, T2::ValOwned), T2::Time, T2::Diff)>,
+        L: FnMut(T1::Key<'_>, &[(T1::Val<'_>, T1::Diff)], &mut Vec<(T2::ValOwned, T2::Diff)>)
+            + 'static,
+        Arranged<G, TraceAgent<T2>>: ArrangementSize;
 }
 
-impl<G, K, V, T1, R> MzReduce<G, K, V, R> for Arranged<G, T1>
+impl<G, T1> MzReduce<G, T1> for Arranged<G, T1>
 where
     G::Timestamp: Lattice + Ord,
     G: Scope,
-    K: Data,
-    V: Data,
-    R: Semigroup,
-    T1: for<'a> TraceReader<
-            Key<'a> = &'a K,
-            KeyOwned = K,
-            Val<'a> = &'a V,
-            Time = G::Timestamp,
-            Diff = R,
-        > + Clone
-        + 'static,
+    T1: TraceReader<Time = G::Timestamp> + Clone + 'static,
+    T1::Diff: differential_dataflow::difference::Semigroup,
 {
+    /// Applies `reduce` to arranged data, and returns an arrangement of output data.
+    fn mz_reduce_abelian<L, T2>(&self, name: &str, logic: L) -> Arranged<G, TraceAgent<T2>>
+    where
+        T2: for<'a> Trace<Key<'a> = T1::Key<'a>, Time = G::Timestamp> + 'static,
+        T2::ValOwned: Data,
+        T2::Diff: Abelian,
+        T2::Batch: Batch,
+        T2::Builder:
+            Builder<Output = T2::Batch, Item = ((T1::KeyOwned, T2::ValOwned), T2::Time, T2::Diff)>,
+        L: FnMut(T1::Key<'_>, &[(T1::Val<'_>, T1::Diff)], &mut Vec<(T2::ValOwned, T2::Diff)>)
+            + 'static,
+        Arranged<G, TraceAgent<T2>>: ArrangementSize,
+    {
+        // Allow access to `reduce_abelian` since we're within Mz's wrapper.
+        #[allow(clippy::disallowed_methods)]
```

Comment:
I think we have to update `clippy.toml` to disallow the new functions. Can do in a separate PR.